features/mbedtls/targets/TARGET_STM/md5_alt.c

```diff
@@ -170,11 +170,12 @@ void mbedtls_md5_finish( mbedtls_md5_context *ctx, unsigned char output[16] )
     if (st_md5_restore_hw_context(ctx) != 1) {
         return; // Return HASH_BUSY timout error here
     }
-    if (ctx->sbuf_len > 0) {
-        if (HAL_HASH_MD5_Accumulate(&ctx->hhash_md5, ctx->sbuf, ctx->sbuf_len) != 0) {
-            return; // Return error code here
-        }
+    /* Last accumulation for extra bytes in sbuf_len */
+    /* This allows the HW flags to be in place in case mbedtls_sha256_update has not been called yet */
+    if (HAL_HASH_MD5_Accumulate(&ctx->hhash_md5, ctx->sbuf, ctx->sbuf_len) != 0) {
```

Comment:
@adustm Thank you for the changes! As far as I see, there's still a redundancy in `xxx_update`: with the current version of `xxx_finish` which always calls `HAL_HASH_MD5_Accumulate`, it should be fine for `xxx_update` to be a no-op on an empty buffer.

Reply:
Hello @hanno-arm 
If I remove this 2nd redundancy, the md5 mbedtls-selftest is crashing, that's why I've kept it.

Reply:
@adustm To be sure we understand each other correctly: As far as I understand the code, it should be fine to remove the code-path for `currentlen == 0` from `xxx_update` - but leaving `xxx_finish` as it is now.

Reply:
Hello @hanno-arm 
Thanks for insisting, I have made more tests. 
The updated version of the code avoids the remaining redundancy and passes md5_selftest successfully.

Kind regards

Reply:
@adustm Thank you for the rework!